### PR TITLE
API Deprecate FormField API

### PR DESCRIPTION
**Pull request by emteknetnz**
Issue https://github.com/silverstripe/silverstripe-framework/issues/11422

### CODE REVIEW

File: src/Forms/DateField.php [hunk 1]
```diff
@@ -101,6 +101,7 @@ class DateField extends TextField
      * to detect invalid values.
      *
      * @var mixed
+     * @deprecated 5.4.0 Will be removed without equivalent functionality to replace it
```
Comment (GuySartorelli) [inline src/Forms/DateField.php]:
```suggestion
     * @deprecated 5.4.0 Use $value instead
```
Presumably if people want to get the value, they should use the value property. Or the getValue() method.
Please apply this to the other `rawValue` deprecation phpdocs as well.

File: src/Forms/FormField.php [hunk 2]
```diff
@@ -1223,9 +1224,11 @@ public function Type()
      * @param bool $result
      * @param Validator $validator
      * @return bool
+     * @deprecated 5.4.0 Will be replaced by the `updateValidate` extension hook
      */
     protected function extendValidationResult(bool $result, Validator $validator): bool
     {
+        Deprecation::noticeWithNoReplacment('5.4.0', 'Will be replaced by the `updateValidate` extension hook');
```
Comment (GuySartorelli) [inline src/Forms/FormField.php]: The method doesn't get replaced with the extension hook - if I'm calling this method in my custom form field I don't now need to implement an extension class, I need to call `extend()` directly.
```suggestion
        Deprecation::noticeWithNoReplacment('5.4.0', 'Use extend("updateValidate") instead');
```

**Review by GuySartorelli — CHANGES_REQUESTED** [review]

File: src/Forms/FormField.php (hunk 2 above)
Comment (GuySartorelli) [reply src/Forms/FormField.php]: Please also update the phpdoc and the changelog

File: src/Forms/DateField.php (hunk 1 above)
Comment (emteknetnz) [reply src/Forms/DateField.php]: I think it's correct as is as $this->value is still getting modified in CMS 5 - as mentioned on https://github.com/silverstripe/silverstripe-framework/pull/11449#discussion_r1839366435 this property has a specific purpose

File: src/Forms/FormField.php (hunk 2 above)
Comment (emteknetnz) [reply src/Forms/FormField.php]: The `updateValidate` hook does not exist in CMS 5, so we cannot give this recommendation.

File: src/Forms/DateField.php (hunk 1 above)
Comment (GuySartorelli) [reply src/Forms/DateField.php]: That comment mentions the past tense, and https://github.com/silverstripe/silverstripe-framework/pull/11449#discussion_r1839364649 says that `getValue()` should return the value "as unmodified as possible". Which I took to mean that the `value` property would now be the raw value, and the various methods for getting the value are where any modification of the value would take place.

Is there no way to get just _the_ value that was set to the field now? I thought that was explicitly the purpose of `getValue()`?

File: src/Forms/FormField.php (hunk 2 above)
Comment (GuySartorelli) [reply src/Forms/FormField.php]: `use extend() directly instead` then. The fact is that you can stop calling this method in 5 immediately - its replacement is the `extend()` method. The change of extension hook method isn't really related to this method being deprecated.

Comment (emteknetnz) [reply src/Forms/FormField.php]: I don't understand how you'd use extend() directly in CMS 5, or at least why you would

The thing being replaced here has a bool + a Validator for params.  The functionality that's replacing it in CMS 6 has a ValidationResult param instead.

Presumably you'd need to implement something sort of convoluted in CMS 5 for this to work ... only to then replace it straight away in CMS 6?

File: src/Forms/DateField.php (hunk 1 above)
Comment (emteknetnz) [reply src/Forms/DateField.php]: In the context of Date/Datetime/TimeField, $this->value will get modified after calling setValue($value). $this->rawValue is explicitly not getting modified at any point. There's no equivalent functionality in CMS 5.

My intention, at least, in CMS 6 at least is to not modify $this->value after setting it so that when validation is run it's running on the value that was set by the developer, though in practice Fields could still alter $this->value.

tl;dr Will be removed without equivalent functionality to replace it :-)

File: src/Forms/FormField.php (hunk 2 above)
Comment (GuySartorelli) [reply src/Forms/FormField.php]: > I don't understand how you'd use extend() directly in CMS 5, or at least why you would

`extendValidationResult()` is ultimately an abstraction around `$this->extend('updateValidationResult', $result, $validator);`

You can just call that directly.

In CMS 5 you'll get a deprecation warning if you're calling `extendValidationResult()`. One way to silence that is to call `extend()` directly instead of calling `extendValidationResult()`.

In CMS 6 that will also be the solution - calling `extend()`, since `extendValidationResult()` will no longer exist. Just you'll be passing a different hook name and different args, which is not usually information we include in deprecation notice messages.

File: src/Forms/DateField.php (hunk 1 above)
Comment (GuySartorelli) [reply src/Forms/DateField.php]: Discussed in person - `$value` is functionally equivalent to `$rawValue` even if the exact value isn't identical

Comment (emteknetnz) [reply src/Forms/DateField.php]: Updated

File: src/Forms/FormField.php (hunk 2 above)
Comment (emteknetnz) [reply src/Forms/FormField.php]: Updated

**Review by GuySartorelli — APPROVED** [review]
LGTM
